tsdb/engine/tsm1/engine.go

```diff
@@ -2206,13 +2277,79 @@ func (e *Engine) compact(wg *sync.WaitGroup) {
 			e.CompactionPlan.Release(level2Groups)
 			e.CompactionPlan.Release(level3Groups)
 			e.CompactionPlan.Release(level4Groups)
+			e.CompactionPlan.Release(level5Groups)
+		}
+	}
+}
+
+func (e *Engine) PlanCompactions() (level1Groups []CompactionGroup,
+	level2Groups []CompactionGroup,
+	level3Groups []CompactionGroup,
+	level4Groups []CompactionGroup,
+	level5Groups []CompactionGroup,
+	level5Aggressive bool) {
+	// Find our compaction plans
+	level1Groups, len1 := e.CompactionPlan.PlanLevel(1)
+	level2Groups, len2 := e.CompactionPlan.PlanLevel(2)
+	level3Groups, len3 := e.CompactionPlan.PlanLevel(3)
+	initialLevellevel4Groups, _ := e.CompactionPlan.Plan(e.LastModified())
+
+	// Some groups in level 4 may contain already optimized files. In these cases, we want
+	// to maintain optimization for the entire group to avoid "going backwards" on the
+	// optimization level. For instance, if an optimized cold shard had back-fill data
+	// added to it, we should maintain the optimization to avoid unoptimizing the bulk of
+	// the shards only to need to reoptimize them later.
+	// In an ideal world, CompactionPlan.Plan and CompactionPlan.PlanOptimize might handle this.
+	level4Groups = make([]CompactionGroup, 0, len(initialLevellevel4Groups))
+	level5Groups = make([]CompactionGroup, 0, len(initialLevellevel4Groups))
+	for _, group := range initialLevellevel4Groups {
+		if isOpt, filename, heur := e.IsGroupOptimized(group); isOpt {
+			e.logger.Info("Promoting full compaction level 4 group to optimized level 5 compaction group because it contains an already optimized TSM file",
+				zap.String("optimized_file", filename), zap.String("heuristic", heur), zap.Strings("files", group))
+			level5Groups = append(level5Groups, group)
+		} else {
+			level4Groups = append(level4Groups, group)
 		}
 	}
+
+	// level5Aggressive indicates if /all/ level 5 compactions should be done at the aggressive points per block, or only those
+	// which have already been done at the aggressive points per block.
+	// If no full compactions are needed, see if an optimize is needed
+	if len(level4Groups) == 0 {
+		plannedLevel5Groups, _, genCount := e.CompactionPlan.PlanOptimize(e.LastModified())
```

Comment:
I believe we return the wrong generation count here because `DefaultPlanner.PlanOptimize` calls `DefaultPlanner.findGenerations(true)`, skipping files in use.  But, at this point,  files we are looking at are already in a plan. See notes in  `TestEnginePlanCompactions`